File: Device/SFBLevelControl.m
/*
 * Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

#import "SFBLevelControl.h"
#import "SFBAudioObject+Internal.h"

#import "SFBCStringForOSType.h"

@implementation SFBLevelControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioLevelControlClassID));
	return [super initWithAudioObjectID:objectID];
}

- (NSNumber *)scalarValue
{
	return [self floatForProperty:kAudioLevelControlPropertyScalarValue inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setScalarValue:(float)value error:(NSError **)error
{
	return [self setFloat:value forProperty:kAudioLevelControlPropertyScalarValue inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSNumber *)decibelValue
{
	return [self floatForProperty:kAudioLevelControlPropertyDecibelValue inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setDecibelValue:(float)value error:(NSError **)error
{
	return [self setFloat:value forProperty:kAudioLevelControlPropertyDecibelValue inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSValue *)decibelRange
{
	return [self audioValueRangeForProperty:kAudioLevelControlPropertyDecibelRange inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)convertToDecibelsFromScalar:(float)scalar error:(NSError **)error
{
	AudioObjectPropertyAddress propertyAddress = {
		.mSelector	= kAudioLevelControlPropertyConvertScalarToDecibels,
		.mScope		= kAudioObjectPropertyScopeGlobal,
		.mElement	= kAudioObjectPropertyElementMaster
	};

	Float32 value = scalar;
	UInt32 dataSize = sizeof(value);
	OSStatus result = AudioObjectGetPropertyData(_objectID, &propertyAddress, 0, NULL, &dataSize, &value);
	if(result != kAudioHardwareNoError) {
		os_log_error(gSFBAudioObjectLog, "AudioObjectGetPropertyData (kAudioLevelControlPropertyConvertScalarToDecibels) failed: %d '%{public}.4s'", result, SFBCStringForOSType(result));
		if(error)
			*error = [NSError errorWithDomain:NSOSStatusErrorDomain code:result userInfo:nil];
		return nil;
	}

	return @(value);
}

- (NSNumber *)convertToScalarFromDecibels:(float)decibels error:(NSError **)error
{
	AudioObjectPropertyAddress propertyAddress = {
		.mSelector	= kAudioLevelControlPropertyConvertDecibelsToScalar,
		.mScope		= kAudioObjectPropertyScopeGlobal,
		.mElement	= kAudioObjectPropertyElementMaster
	};

	Float32 value = decibels;
	UInt32 dataSize = sizeof(value);
	OSStatus result = AudioObjectGetPropertyData(_objectID, &propertyAddress, 0, NULL, &dataSize, &value);
	if(result != kAudioHardwareNoError) {
		os_log_error(gSFBAudioObjectLog, "AudioObjectGetPropertyData (kAudioLevelControlPropertyConvertDecibelsToScalar) failed: %d '%{public}.4s'", result, SFBCStringForOSType(result));
		if(error)
			*error = [NSError errorWithDomain:NSOSStatusErrorDomain code:result userInfo:nil];
		return nil;
	}

	return @(value);
}

@end

@implementation SFBVolumeControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioVolumeControlClassID));
	return [super initWithAudioObjectID:objectID];
}

@end

@implementation SFBLFEVolumeControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioLFEVolumeControlClassID));
	return [super initWithAudioObjectID:objectID];
}

@end

File: Device/SFBLevelControl.swift
/*
* Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
* See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
*/

import Foundation

extension LevelControl {
	/// Returns the control's scalar value
	/// - note: This corresponds to `kAudioLevelControlPropertyScalarValue`
	/// - throws: An error if the property could not be retrieved
	public func scalarValue() throws -> Float {
		return try getProperty(.levelControlScalarValue)
	}

	/// Sets the control's scalar value
	/// - note: This corresponds to `kAudioLevelControlPropertyScalarValue`
	/// - parameter value: The desired property value
	/// - throws: An error if the property could not be set
	public func setScalarValue(_ value: Float) throws {
		try setProperty(.levelControlScalarValue, value)
	}

	/// Returns the control's decibel value
	/// - note: This corresponds to `kAudioLevelControlPropertyDecibelValue`
	/// - throws: An error if the property could not be retrieved
	public func decibelValue() throws -> Float {
		return try getProperty(.levelControlDecibelValue)
	}

	/// Sets the control's decibel value
	/// - note: This corresponds to `kAudioLevelControlPropertyDecibelValue`
	/// - parameter value: The desired property value
	/// - throws: An error if the property could not be set
	public func setDecibelValue(_ value: Float) throws {
		try setProperty(.levelControlDecibelValue, value)
	}

	/// Returns the decibel range
	/// - note: This corresponds to `kAudioLevelControlPropertyDecibelRange`
	/// - throws: An error if the property could not be retrieved
	public func decibelRange() throws -> AudioValueRange {
		return try getProperty(.levelControlDecibelRange)
	}

	/// Converts `scalar` to decibels and returns the converted value
	/// - note: This corresponds to `kAudioLevelControlPropertyConvertScalarToDecibels`
	/// - parameter scalar: The value to convert
	/// - throws: An error if the property could not be set
	public func convertToDecibels(_ scalar: Float) throws -> Float {
		return try __convertToDecibels(fromScalar: scalar).floatValue
	}

	/// Converts `decibels` to scalar and returns the converted value
	/// - note: This corresponds to `kAudioLevelControlPropertyConvertDecibelsToScalar`
	/// - parameter decibels: The value to convert
	/// - throws: An error if the property could not be set
	public func convertToScalar(_ decibels: Float) throws -> Float {
		return try __convertToScalar(fromDecibels: decibels).floatValue
	}
}
